tools/bas2prg.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python3
+
+# Copyright (c) 2019, Frank Buss
+# All rights reserved.
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+import argparse
+import os
+import codecs
+
+# parse arguments
+parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+    description='Converts a BAS files (text format) to PRGs (X16 native format) with the emulator.')
+parser.add_argument('emulator', help='path to the x16emu program')
+parser.add_argument('input', help='the input directory with BAS files')
+parser.add_argument('output', help='the PRG output directory')
+args = parser.parse_args()
+
+for f in os.listdir(args.input):
+    if f.endswith('.bas'):
+        basFilename = args.input + "/" + f
+        print("converting file %s..." % basFilename)
+        prg = f.upper()[:-4] + ".PRG"
+        prgFilename = args.output + "/" + prg
+        prg = prg + ".TMP"
+
+        # read BASIC text file
+        with codecs.open(basFilename, "r", "iso-8859-1") as input:
```

Comment:
The emulator only really pushes bytes into the keyboard buffer, so ISO 8859-1 sounds like it is meant as a way to tell Python to just take 8 bits per character and not touch the encoding? Does this mean you will write the exact same bytes back out again?

Can you instead read the data into a ByteArray? That will make it much clearer that you are not touching the encoding.

Reply:
There was an error while reading it. I though with the encoding this would be fixed, but it changed one file. Now it is binary and reading and writing works. So this should be good now.
But now the emulator stops with the file layer-demo.bas, at strange print at line 9930. This should be fixed in the emulator, or ask the author what it means and maybe replaced with a chr$, then it would work.